### ISSUE TITLE
Feature request: dictionaries scope

### ISSUE THREAD

**Issue by filimonov** [ClickHouse/ClickHouse#4201]
It's quite common in staging environment to have multiple databases, one for each branch / stage.

Like 
```
 database_rc 
   - table1
   - table2

 database_master
   - table1
   - table2 

 database_testing
   - table1
   - table2

 database_feature_branch_xxx
   - table1
   - table2
   - table3a
 ```

It's easy to do in ClickHouse and (for example) in MySQL separately. 

But if you want each database in ClickHouse to use the dictionary connected to MySQL of the corresponding stage - it's not possible, as dictionaries are in global namespace in ClickHouse.

So the proposition is - to create a scope for dictionaries, with the similar configuration syntax as <allow_databases>. By default, the global scope would be used. If the scope is defined, then that dictionary is registered in corresponding database namespace, otherwise - in the global namespace. If two dictionaries with the same name are registered in the same namespace - then exception happen. If the same dictionary is defined in DB scope and global scope - more specialized (i.e. from database scope) should be used.

That would allow creating the dictionary with the same name and different definition in different databases.

**Comment by filimonov** [comment]
Also. Something like SYSTEM UNLOAD DICTIONARY would be cool (to free the memory used by dictionary, and put it to 'init on demand' state, like for lazy load case)

**Comment by s-mx** [comment]
Hello @filimonov . Very soon we will merge this PR https://github.com/yandex/ClickHouse/pull/4178 which allows create dictionaries in SQL-like queries and it definition will be attached to database with engine Dictionary.
So user can create, for example, database a with next query
`CREATE DATABASE a engine = Dictionary`
and after that user can define dictionary with query like next
`CREATE DICTIONARY a.dict (field1 Type1, ...) PRIMERY KEY key1 engine = Dictionary(SOURCE(...))`
It seems, that's what you need.

**Comment by filimonov** [comment]
> It seems, that's what you need.
Not exactly. WHile that feature (DDL for dicts) is really cool, i was telling about scope of usage of dictionaries, not the scope of definition.

I.e. `dictGetString('dictionary',...)` called while we are inside `database_rc` should use one dictionary (attached to `rc` brach), and the same call `dictGetString('dictionary',...)` (with the same dictionary name) but used while we are inside `database_devel` should use another dictionary (attached to `devel` branch).

**Comment by filimonov** [comment]
That is implemented in  #7360 and further PRs, avaliable only with DDL dictionary syntax in clickhouse versions >= 20.1

**Comment by joeyutong** [comment]
> Also. Something like SYSTEM UNLOAD DICTIONARY would be cool (to free the memory used by dictionary, and put it to 'init on demand' state, like for lazy load case)

Is there a way to do something like this now? @filimonov
